content/blog/2020/01/2020-01-23-ui-plugins.adoc

```diff
@@ -0,0 +1,701 @@
+by using the new tag `svg-icon` in your jelly view:
+
+[source,xml,linenums]
+.index.jelly
+----
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:fa="/font-awesome">
+
+  [...]
+  <fa:svg-icon name="check-double" class="no-issues-banner"/>
+  [...]
+
+</j:jelly>
+----
+
+If you are generating views using Java code, then you also can use the class `SvgTag` to generate the
+HTML markup for such an icon.
+
+[#boostrap-grid]
+=== Grid layout
+
+The first thing to decide is, which elements should be shown on a plugin page and how much space each element
+should occupy. Typically, all visible components are mapped on the available space using a simple grid.
+In a Jenkins view we have a fixed header and footer and a navigation bar on the left
+(20 percent of the horizontal space). The rest of a screen can be used by
+a details view. In order to simplify the distribution of elements in that remaining space we use
+https://getbootstrap.com/docs/4.4/layout/grid/[Bootstrap's grid system].
+
+.Jenkins layout with a details view that contains a grid system
+[#img-grid]
+image::grid.png[Grid layout in Jenkins]
+
+That means, a view is split into 12 columns and and arbitrary number of rows. This grid system is simple to use
+(but complex enough to also support fancy screen layouts) - I won't go into
+details here, please refer to the https://getbootstrap.com/docs/4.4/layout/grid/[Bootstrap documentation]
+for details.
+
+For the forensics detail view we use a simple grid of two rows and two columns. Since the number of columns always is 12
+we need to create two "fat" columns that fill 6 of the standard columns.
+In order to create such a view in our
+plugin we need to create a view given as a jelly file and a corresponding Java view model object. A view with this layout
+is shown in the following snippet:
+
+[source,xml,linenums]
+.index.jelly
+----
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:bs="/bootstrap">
+
+  <bs:layout title="${it.displayName}" norefresh="true"> <1>
+    <st:include it="${it.owner}" page="sidepanel.jelly"/>
+    <l:main-panel>
+      <st:adjunct includes="io.jenkins.plugins.bootstrap4"/> <2>
+      <div class="fluid-container"> <3>
+
+        <div class="row py-3"> <4>
+          <div class="col-6"> <5>
+            Content of column 1 in row 1
+          </div>
+          <div class="col-6"> <6>
+            Content of column 2 in row 1
+          </div>
+        </div>
+
+        <div class="row py-3"> <7>
+          <div class="col"> <8>
+            Content of row 2
+          </div>
+        </div>
+
+      </div>
+    </l:main-panel>
+  </bs:layout>
+</j:jelly>
+----
+<1> Use a custom layout based on Bootstrap: since Jenkins core contains an old version of Bootstrap,
+we need to replace the standard layout.jelly file.
```

Comment:
I think encouraging plugin developers to use the custom layout.jelly file present problems because it 100% will break and clash with the new UI. Basically, plugin pages using this layout will be rendered with the old UI. 

If the main issue is that the bootstrap 4 styles clash with the current bootstrap grid, the solution is for the BS plugin to provide a reduced subset of bootstrap, with no grid and no basic styles. Bootstrap's modularity allows this. The grid system interface is basically the same, and the core grid version could be upgraded.

My other question is, can the `<bs:layout>` helper be later updated to just load the the bootstrap scripts and not replacing the layout?

Reply:
> If the main issue is that the bootstrap 4 styles clash with the current bootstrap grid, the solution is for the BS plugin to provide a reduced subset of bootstrap, with no grid and no basic styles. Bootstrap's modularity allows this. 

Yes, that would make sense.  Actually, my JS build experience is very limited, maybe you can give me a helping hand here. What I actually would like to have is something like:
- Have the version of Bootstrap defined in the pom (or somewhere else)
- Define the set of Bootstrap features to use in a configuration file
- Create the js/css files automatically without human interaction.
- Deploy the new artifact as plugin

> The grid system interface is basically the same, and the core grid version could be upgraded.

Well, I did not try to use the BS3 grid. I assume that the breakpoints are the same as in BS4, this is not sufficient for an app that mostly runs on large displays. Are you intending to replace the old BS3 grid with the new version in the near future?

> My other question is, can the <bs:layout> helper be later updated to just load the the bootstrap scripts and not replacing the layout?

Yes, I can move the layout tag back to the warnings plugin so no one else will use it. Then the warnings plugin will remain working with the current set of breakpoints. And other plugins are not affected.

Reply:
On the topic of users being able to generate a customized version of the plugin, I don't know how complex that would be tbh. I'd think it would be enough if the plugins offers modularity when it comes to importing the assets.

A quick idea: maybe the plugin could have adjuncts for each UI item (cards, scrollspy, tables, etc) and load them under 2 conditions:

- Allow plugin authors to load them manually
- Load them automatically once the component's jelly helpers are used. For example, using the `<bs:card>` tag would automatically load the card assets. 

> I assume that the breakpoints are the same as in BS4, this is not sufficient for an app that mostly runs on large displays. Are you intending to replace the old BS3 grid with the new version in the near future?

I just realized that I was mistaken and the grids are not actually compatible. For some weird reason Jenkins uses a 24 column grid :| 
Anyway, the bootstrap grid has a breakpoint for screens over 1800px, the `.col-xlg-N` classes.

If that's not enough, it would be really easy to add another breakpoint to the responsive-grid.css file on a small PR (maybe you could even test it plugin-side). A example of a possible 2200px breakpoint could look like: https://gist.github.com/fqueiruga/f9b86ea415be70802bf7f475c8577526 (px dimensions chosen at random)

Reply:
> On the topic of users being able to generate a customized version of the plugin, I don't know how complex that would be tbh. 

Seems that I did explain that part wrongly. What I wanted to say: I want *this* bootstrap plugin to be simple to maintain, i.e. everything should be automated. Currently I need to download the version, modify the sass to change the grid, and then rebuild. 

> Anyway, the bootstrap grid has a breakpoint for screens over 1800px, the .col-xlg-N classes.

From https://github.com/twbs/bootstrap/blob/master/scss/_variables.scss#L290 it looks like the maximum grid value is 1200px. That is the reason why I introduced some different ones.
Or do you mean the existing Jenkins grid? I can patch that as well if that makes more sense for you. In the long run it would make more sense if we would replace the BS3 grid in Jenkins core as you also suggested...

Reply:
> I want this bootstrap plugin to be simple to maintain, i.e. everything should be automated. Currently I need to download the version, modify the sass to change the grid, and then rebuild.
I think I misunderstood the workflow then. I thought one could just add the proper imports on this file once and then no more updates until bootstrap 5 https://github.com/jenkinsci/bootstrap4-api-plugin/blob/master/etc/custom.scss.

On the topic of the grid, you are right, I meant the existing Jenkins grid. It goes up to 1800px (see https://github.com/jenkinsci/jenkins/blob/master/war/src/main/webapp/css/responsive-grid.css#L1287). I agree that in the long run we should probably have a proper grid. We could even deprecate the current one.

Reply:
@fqueiruga are your concerns well enough addressed that this can be posted?